### PR TITLE
Refactor maintenance commands

### PR DESCRIPTION
**Pull request by zzaakiirr**
### What does this PR do?

This PR refactors code related to maintenance commands:
1. New `MaintenanceMode` class is introduced to follow DRY principle in `Commands::Maintenance`, `Commands::MaintenanceOff` and `Commands::MaintenanceOn`
2. Related tests are refactored - added usage of `shared_examples`

### Note

Triggered [slow rspec job](https://github.com/shakacode/control-plane-flow/actions/runs/10005921322/job/27657561104) manually, maintenance command related specs run successfully

## Summary by CodeRabbit

- **New Features**
  - Introduced a new `MaintenanceMode` class for managing application maintenance states, allowing for more streamlined enabling and disabling of maintenance mode.

- **Improvements**
  - Simplified the `call` methods in `Maintenance`, `MaintenanceOff`, and `MaintenanceOn` classes, focusing on maintenance mode status without complex logic.
  - Enhanced test structure for maintenance commands through shared examples, improving maintainability and reducing redundancy.

- **Chores**
  - Updated the test suite to dynamically load support files, simplifying code management.

### CODE REVIEW

**Comment by zzaakiirr** [comment]
@coderabbitai review

File: spec/support/shared_examples/maintenance_commands.rb [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    it "raises error" do
+      result = run_cpflow_command(command, "-a", app)
+
+      expect(result[:status]).not_to eq(0)
+      expect(result[:stderr]).to include("Can't find workload 'maintenance'")
+    end
+  end
+end
+
+RSpec.shared_examples "switch maintenance mode command" do |action:|
```
Comment (rafaelgomesxyz) [inline spec/support/shared_examples/maintenance_commands.rb:29]: If we're following the same pattern as the ones above, this should be:

```suggestion
RSpec.shared_examples "switches maintenance mode command" do |action:|
```

File: spec/support/shared_examples/maintenance_commands.rb [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+  include_examples "validates domain existence", command: "maintenance:#{action}"
+  include_examples "validates maintenance workload existence", command: "maintenance:#{action}"
+
+  context "when maintenance workload exists" do
+    let!(:app) { dummy_test_app("full", create_if_not_exists: true) }
+
+    let(:command) { "maintenance:#{action}" }
+    let(:opposite_command) { "maintenance:#{enable?(action) ? 'off' : 'on'}" }
+    let(:state_switched_text) { enable?(action) ? "enabled" : "disabled" }
```
Comment (rafaelgomesxyz) [inline spec/support/shared_examples/maintenance_commands.rb:38]: Could be just `state`.

**Review by rafaelgomesxyz — CHANGES_REQUESTED** [review]
Also, let's not run all slow specs on a PR level, unless the PR changes something that affects many different commands, because they're very slow and consume our credits. They're intended to be run only every once in a while in the `main` branch.

Are you able to run specs locally? Because if so, you could run the individual specs locally and post a screenshot of the run in the PR description. If not, I can run them myself to verify for now.

Ideally, we should have a way to run only modified slow specs on a PR level, but I don't know how complex that would be to implement.

File: lib/command/maintenance_on.rb [hunk 3]
```diff
@@ -15,48 +15,14 @@ class MaintenanceOn < Base
...
+      maintenance_mode.enable!
+    end
 
-      # Stop all other workloads
-      run_cpflow_command("ps:stop", "-a", config.app, "--wait")
+    private
 
-      progress.puts("\nMaintenance mode enabled for app '#{config.app}'.")
+    def maintenance_mode
+      @maintenance_mode ||= ::MaintenanceMode.new(self)
```
[the remark this answers is not in the document]
Comment (zzaakiirr) [reply lib/command/maintenance_on.rb]: No, it's not needed - removed

File: lib/core/maintenance_mode.rb [hunk 4]
```diff
@@ -0,0 +1,107 @@
+# frozen_string_literal: true
+
+require "forwardable"
```
[the remark this answers is not in the document]
Comment (zzaakiirr) [reply lib/core/maintenance_mode.rb:3]: > Just an observation for an improvement for a separate PR: we have DoctorService and TemplateParser which would both benefit from the Forwardable logic as well.

👍 I'll create a separate PR for this

File: lib/core/maintenance_mode.rb [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+    raise "Can't find domain. " \
+          "Maintenance mode is only supported for domains that use path based routing mode " \
+          "and have a route configured for the prefix '/' on either port 80 or 443."
+  end
+
+  def validate_maintenance_workload_exists!
+    cp.fetch_workload!(maintenance_workload)
+  end
+
+  def run_all_workloads(start_or_stop)
```
[the remark this answers is not in the document]
Comment (zzaakiirr) [reply lib/core/maintenance_mode.rb]: Couldn't come up with better naming and used `start_or_stop_all_workloads `

File: spec/spec_helper.rb [hunk 6]
```diff
@@ -28,10 +28,8 @@
 end
 
 require_relative "../lib/cpflow"
-require_relative "support/dummy_app_setup"
-require_relative "support/command_helpers"
-require_relative "support/date_time_helpers"
-require_relative "support/stub_env"
+
+Dir["./spec/support/**/*.rb"].sort.each { |f| require f }
```
[the remark this answers is not in the document]
Comment (zzaakiirr) [reply spec/spec_helper.rb]: You are right, better to use `__dir__`

**Comment by zzaakiirr** [comment]
> Great job overall! 💯

🙇 

> Also, let's not run all slow specs on a PR level, unless the PR changes something that affects many different commands, because they're very slow and consume our credits. They're intended to be run only every once in a while in the `main` branch.

Sorry, didn't know that - won't run all slow specs in the future

> Are you able to run specs locally? Because if so, you could run the individual specs locally and post a screenshot of the run in the PR description. If not, I can run them myself to verify for now.

Some specs fail locally - `spec/command/maintenance_off_spec.rb` and `spec/command/maintenance_on_spec.rb` for example. I'll provide screenshots in the future if it will be possible

> Ideally, we should have a way to run only modified slow specs on a PR level, but I don't know how complex that would be to implement.

I'll look into it 👍

**Review by rafaelgomesxyz — APPROVED** [review]
> Sorry, didn't know that - won't run all slow specs in the future

No problem. It's been a while since we ran it anyway, so it's good that you ran it.

> Some specs fail locally - spec/command/maintenance_off_spec.rb and spec/command/maintenance_on_spec.rb for example. I'll provide screenshots in the future if it will be possible

I realize that you don't have access to the Control Plane org, so any specs that rely on network requests (which is most of the `spec/command` ones) will unfortunately fail for you locally.

I ran them here and they're all good:

maintenance-specs

> I'll look into it 👍

If it seems too complex, don't spend too much time on it, I can just run the specs manually when needed.

I guess a simple alternative would be to add a GitHub workflow that can be triggered manually, and accepts args for `bundle exec rspec`. That way we could selectively run only the specs we want on CI.

**Pull request merged**